table/block_based/block_based_table_builder.cc

```diff
@@ -764,31 +1017,119 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
         size_t pad_bytes =
             (r->alignment - ((block_contents.size() + kBlockTrailerSize) &
                              (r->alignment - 1))) &
             (r->alignment - 1);
-        r->status = r->file->Pad(pad_bytes);
-        if (r->status.ok()) {
+        SetStatusAtom(r->file->Pad(pad_bytes));
+        if (ok()) {
           r->offset += pad_bytes;
         }
       }
     }
   }
 }
 
-Status BlockBasedTableBuilder::status() const { return rep_->status; }
+void BlockBasedTableBuilder::BGWorkWriteRawBlock() {
+  Rep* r = rep_;
+  ParallelCompressionRep::BlockRepSlot* slot;
+  ParallelCompressionRep::BlockRep* block_rep;
+  while (r->pc_rep->write_queue.pop(slot)) {
+    slot->Take(block_rep);
+    if (!block_rep->status.ok()) {
+      SetStatusAtom(block_rep->status);
+      break;
+    }
+
+    for (size_t i = 0; i < block_rep->keys->Size(); i++) {
+      auto& key = (*block_rep->keys)[i];
+      if (r->filter_builder != nullptr) {
+        size_t ts_sz =
+            r->internal_comparator.user_comparator()->timestamp_size();
+        r->filter_builder->Add(ExtractUserKeyAndStripTimestamp(key, ts_sz));
+      }
+      r->index_builder->OnKeyAdded(key);
+    }
+
+    r->pc_rep->raw_bytes_curr_block = block_rep->data->size();
+    WriteRawBlock(block_rep->contents, block_rep->compression_type,
+                  &r->pending_handle, true /* is_data_block*/);
+    if (!ok()) {
+      break;
+    }
+
+    if (r->pc_rep->first_block) {
+      std::unique_lock<std::mutex> lock(r->pc_rep->first_block_mutex);
+      r->pc_rep->first_block = false;
+      r->pc_rep->first_block_cond.notify_one();
+    }
+
+    if (r->filter_builder != nullptr) {
+      r->filter_builder->StartBlock(r->offset);
+    }
+    r->props.data_size = r->offset;
+    ++r->props.num_data_blocks;
+
+    if (block_rep->first_key_in_next_block == nullptr) {
+      r->index_builder->AddIndexEntry(&(block_rep->keys->Back()), nullptr,
+                                      r->pending_handle);
+    } else {
+      Slice first_key_in_next_block =
+          Slice(*block_rep->first_key_in_next_block);
+      r->index_builder->AddIndexEntry(&(block_rep->keys->Back()),
+                                      &first_key_in_next_block,
+                                      r->pending_handle);
+    }
+    block_rep->compressed_data->clear();
+    r->pc_rep->block_rep_pool.push(block_rep);
+  }
+}
+
+Status BlockBasedTableBuilder::status() const {
+  if (rep_->compression_opts.parallel_threads > 1) {
+    std::unique_lock<std::mutex> lock(rep_->status_mutex);
+    return rep_->status;
+  } else {
+    return rep_->status;
+  }
+}
+
+template <typename T>
+void BlockBasedTableBuilder::SetStatusAtom(T&& status) {
+  if (rep_->compression_opts.parallel_threads > 1) {
+    std::unique_lock<std::mutex> lock(rep_->status_mutex);
+    rep_->status = std::forward<T>(status);
+  } else {
+    rep_->status = std::forward<T>(status);
```

Comment:
I think it over complicate the logic. If the only time we set status is when it is not OK, it is a rare operation and the performance for that code path is less important. I would suggest we do a simply copying in this case.

Reply:
Replaced that with simple copy. Now status's produced in `WriteRawBlocks()` are set to global by `SetStatusAtom()`  only in non-OK cases to avoid copy overhead.